File: 2601-prime-subtraction-operation/2601-prime-subtraction-operation.cpp
class Solution {
private:
    vector<bool> calculatePrimeNumbers(int n){
        vector<bool> prime(n+1,true);
        prime[0] = prime[1] = false;
        for(int i=2; i<=n; i++){
            if(prime[i] == true){
                for(int j=i*i; j<=n; j+=i){
                    prime[j] = false;
                }
            }
        }
        return prime;
    }      
public:
    bool primeSubOperation(vector<int>& nums) {
        vector<bool> primeFlag = calculatePrimeNumbers(1005);
        
        vector<int> prime;        
        for(int i=0; i<primeFlag.size(); i++){
            if(primeFlag[i]){
                prime.push_back(i);
            }
        }
        
        int n = nums.size();
        
        vector<int> newNums(n,-1);
        
        newNums[n-1] = nums[n-1];
     
        for(int i=n-2; i>=0; i--){
            if(nums[i] < newNums[i+1]){
                newNums[i] = nums[i];
                continue;
            }            
            for(int p : prime){
                if(p >= nums[i]){
                    newNums[i] = nums[i];
                    break;
                }
                if(nums[i]-p < newNums[i+1]){
                    newNums[i] = nums[i]-p;
                    break;
                }
            }
        }
        
        for(int i=1; i<n; i++){
            if(newNums[i] <= newNums[i-1]){
                return false;
            }
        }
        
        return true;
    }
};

File: 2601-prime-subtraction-operation/2601-prime-subtraction-operation.java
class Solution {
    private List<Integer> primeNumbers;

    public Solution() {
        sieveAlgo();
    }

    public void sieveAlgo() {
        int n = 10000;

        boolean[] prime = new boolean[n + 1];

        Arrays.fill(prime, true);

        prime[0] = prime[1] = false;

        for(int i = 2; (i*i) <= n; i++) {
            if(prime[i]) {
                for(int j = i * i; j <= n; j += i) {
                    prime[j] = false;
                }
            }
        }

        primeNumbers = new ArrayList<>();

        for(int i = 2; i <= n; i++) {
            if(prime[i]) {
                primeNumbers.add(i);
            }
        }
    }

    public boolean primeSubOperation(int[] nums) {
        int n = nums.length;
        
        for(int i = n-2; i >= 0; i--) {
            if(nums[i] < nums[i + 1]) {
                continue;
            }
            
            boolean flag = false;
            
            for(int j = 0; 
                    j < primeNumbers.size() && primeNumbers.get(j) < nums[i]; 
                        j++) {
                
                if((nums[i] - primeNumbers.get(j)) < nums[i + 1]) {
                    nums[i] = nums[i] - primeNumbers.get(j);
                    flag = true;
                    break;
                }
            }
            
            if(flag == false) {
                return false;
            }
        }
        
        return true;
    }
}